### PR TITLE
[kmac] Rework Keccak/SHA3 core and PRNG for SCA hardening

### PR DESCRIPTION
**Pull request by vogelpi**
This PR contains the main results of the KMAC SCA hardening work @vrozic , @andreaskurth and I have been doing over the last couple of weeks. I already briefly talked about this with @eunchan. We propose to merge these changes to substantially reduce both 1st- and 2nd order SCA leakage in KMAC.

The changes can be grouped as follows (I recommend reviewing the individual commits):
- Changes to the `prim_dom_and_2share` primitive (the DOM multiplier): allow optional full pipelining, output intermediate results for remasking other instances
- Changes to the masked Keccak/SHA3 core: enable full pipelining, reduce data width to reduce gate count and PRNG requirements, rework input/output and randomness muxing
- Changes to the PRNG: Add a full-width PRNG consisting of 25 32-bit LFSRs to avoid stacking and concatenating of subsequent LFSR outputs problematic for SCA, avoid buffering in reseeding paths to reduce gate count

Remarks:
1. I've ran a full KMAC regression suite and did get the same failures as we have currently in the nightly regressions. This PR shouldn't require additional DV effort.
1. This PR shouldn't require us to do another, full D2S review. The actual changes to the RTL are smaller than it may seem and we've followed the same approaches as in AES. Changes requested during the code review of this PR will of course be implemented. 
1. This PR leads to a small area reduction (-3.3%, now 186.4 kGE), and a performance improvement when masking is enabled (latency per Keccak round reduced by -40%, 4 instead of 7 clock cycles).
1. Documentation updates are currently missing, I will do those once it's clear if we indeed want to move forward with this.
1. On top of this PR, two more simple changes will follow that are not yet public.

For evaluation results, see [Slide 21 onward](https://docs.google.com/presentation/d/1qMRfFGsjk9t-ZbYpUrikwiJ4WpEXm6VOAShwkROsec4/edit#slide=id.g121c975a7b5_0_0).

### CODE REVIEW

**Comment by vogelpi** [comment]
FYI @GregAC

**Comment by tjaychen** [comment]
@vogelpi do we see an area reduction even with the 25x LFSRs..?

**Comment by vogelpi** [comment]
> @vogelpi do we see an area reduction even with the 25x LFSRs..?

Yes, the area numbers include include everything. The savings in the Keccak core (no input muxing, less output muxing, less comb logic) are more than 8 kGE. As for the PRNG, before we used 2x 64 FFs (double LFSR), 64 FF (EDN buffer), 64 FF (SW buffer), 320 FF (entropy storage) which amounts to 576 FFs. The new PRNG just uses 800 FFs for the 25 32-bit LFSRs, there is no buffering in the reseeding paths anymore.

**Comment by tjaychen** [comment]
> > @vogelpi do we see an area reduction even with the 25x LFSRs..?
> 
> Yes, the area numbers include everything. The savings in the Keccak core (no input muxing, less output muxing, less comb logic) are more than 8 kGE. As for the PRNG, before we used 2x 64 FFs (double LFSR), 64 FF (EDN buffer), 64 FF (SW buffer), 320 FF (entropy storage) which amounts to 576 FFs. The new PRNG just uses 800 FFs for the 25 32-bit LFSRs, there is no buffering in the reseeding paths anymore.

ah thanks for explaining, i'm still looking through the PR, maybe i'll have a better understanding after.

**Comment by vogelpi** [comment]
> ah thanks for explaining, i'm still looking through the PR, maybe i'll have a better understanding after.

You're welcome. I would look at the individual commits separately. It's probably easier to understand. The commit messages try to explain things in addition to the comments in the code.

File: hw/ip/kmac/rtl/keccak_2share.sv
Comment (vogelpi) [reply hw/ip/kmac/rtl/keccak_2share.sv:168]: Yes exactly!

File: hw/ip/kmac/doc/_index.md
Comment (tjaychen) [inline hw/ip/kmac/doc/_index.md:262]: wait is this duplication still true? I thought because DOM is half size now, the 800b is exactly what we need, did I misunderstand?

File: hw/ip/kmac/rtl/keccak_round.sv
Comment (vogelpi) [reply hw/ip/kmac/rtl/keccak_round.sv:258]: This is correct, too. We end up in that situation once per round. Meaning without this, you would get additional leakage on every intermediate round state.

File: hw/ip/kmac/rtl/kmac.sv
Comment (tjaychen) [inline hw/ip/kmac/rtl/kmac.sv:1098]: this can be done as a separate PR, but do you think it makes sense that we enhance `prim_edn_req` so that if the widths match, it directly outputs instead of going through a packing FIFO? I mainly want to do that so we can centralize the assertion below into one place in case it ever chagnes.

File: hw/ip/kmac/doc/_index.md
Comment (vogelpi) [reply hw/ip/kmac/doc/_index.md:262]: Ah sorry, I should have removed this. I still need to update the documentation and will do this once there is agreement we want to move forward with this PR.

File: hw/ip/kmac/rtl/keccak_round.sv
Comment (tjaychen) [reply hw/ip/kmac/rtl/keccak_round.sv:258]: wait sorry, just so i understand. When you say additional leakage, you mean without moving directly on the randomness, we would end up using stale randomness with another operation?

File: hw/ip/kmac/rtl/kmac_entropy.sv
Comment (tjaychen) [inline hw/ip/kmac/rtl/kmac_entropy.sv:323]: i wonder if we need to add some hooks for hardware to trigger this.
If I remember correctly, the software route isn't usable once we're beyond a certain state.

File: hw/ip/kmac/rtl/kmac_entropy.sv
Comment (tjaychen) [inline hw/ip/kmac/rtl/kmac_entropy.sv:335]: for done, should we do `&lfsr_seed_en`. just wondering if that means they're all enabled.

**Comment by vogelpi** [comment]
> @vogelpi one related question, in your evaluation, is data masking typically needed for kmac? One theory we had before was that the part of kmac that operated on the key generally came before the data, so the thinking was that once the key was processed, the data that comes later in theory should not show leakage on the key. Is that actually the case?

I would assume masking of the Keccak core is not strictly required for the part after absorbing the key. However, we did not systematically analyze this. Note that in addition to the masking inside the Keccak core, KMAC also features the option to mask the key/data upon loading into the Keccak state. This was always enabled in our experiments.

Anyway, with this PR the only reason to disable the masking of the Keccak core (fast_entropy mode) would be to save power. But performance-wise there is no longer a difference.

File: hw/ip/kmac/rtl/kmac_entropy.sv
Comment (vogelpi) [reply hw/ip/kmac/rtl/kmac_entropy.sv:335]: Note that we only get 32 bits of entropy from EDN/SW at a time and thus will never set all bits of `lfsr_seed_en` to 1 at the same time.

Comment (tjaychen) [reply hw/ip/kmac/rtl/kmac_entropy.sv:335]: ah sorry i misinterpreted this to mean done for all seeds

File: hw/ip/kmac/rtl/keccak_2share.sv
Comment (tjaychen) [reply hw/ip/kmac/rtl/keccak_2share.sv:168]: o i see. so now every 4 cycles, we need 2x800b of fresh entropy, and 2 sets of intermediate results.

File: hw/ip/kmac/rtl/keccak_round.sv
Comment (tjaychen) [reply hw/ip/kmac/rtl/keccak_round.sv:258]: sorry i think this is the part i'm confused about, instead of relying on an early indication of rand, why not just hold off on changing the inputs into keccak until it is ready? As I understand that has to be done already.

btw, i'm not suggesting we change anything here, just trying to understand.

File: hw/ip/kmac/rtl/kmac_entropy.sv
Comment (tjaychen) [reply hw/ip/kmac/rtl/kmac_entropy.sv:323]: @eunchan let me know if i'm wrong.

File: hw/ip/kmac/rtl/kmac_entropy.sv
Comment (vogelpi) [reply hw/ip/kmac/rtl/kmac_entropy.sv:335]: `lfsr_seed_done` is indeed indicating that we're done with all seeds. Basically, we wait for `seed_idx_q` to point to the last index and when we then get any other `lfsr_seed_en`, we know we are done.

**Comment by tjaychen** [comment]
> > @vogelpi one related question, in your evaluation, is data masking typically needed for kmac? One theory we had before was that the part of kmac that operated on the key generally came before the data, so the thinking was that once the key was processed, the data that comes later in theory should not show leakage on the key. Is that actually the case?
> 
> I would assume masking of the Keccak core is not strictly required for the part after absorbing the key. However, we did not systematically analyze this. Note that in addition to the masking inside the Keccak core, KMAC also features the option to mask the key/data upon loading into the Keccak state. This was always enabled in our experiments.
> 
> Anyway, with this PR the only reason to disable the masking of the Keccak core (fast_entropy mode) would be to save power. But performance-wise there is no longer a difference.
sounds good!

**Comment by tjaychen** [comment]
thanks @vogelpi !
this looks really impressive! I think I generally understand what each commit is trying to do, the main part I'm still not following (and this is more of a general confusing) is when entropy is supplied vs when data is supplied (same question I raised above).

Otherwise this looks good to me. 

Last note, I almost think all your commit messages are worth becoming its own doc, explaining the general principles of hardening ( a lot of the intermediate re-use and PRNG exercise I believe is also common with AES).  What do you think?

File: hw/ip/kmac/rtl/kmac_entropy.sv
Comment (tjaychen) [reply hw/ip/kmac/rtl/kmac_entropy.sv:335]: sounds good. So as i understand `lfsr_seed_en` is never meant to be all set, so it's more a reliance on the `seed_idx_q`. 
Do you think this means `seed_idx_q` should be a `prim_count` of sorts? :) I guess an attack there could potentially skip some seed updates?

File: hw/ip/kmac/rtl/keccak_round.sv
Comment (vogelpi) [reply hw/ip/kmac/rtl/keccak_round.sv:258]: That's exactly what we're doing here: if `rand_early_i` is high, this means we will have fresh randomness in the next cycle. Only then we set the write enable for the state register `update_storage`. The new state (result of Phase1) then becomes visible to the DOM multipliers together with the new randomness in the next cycle.

File: hw/ip/kmac/rtl/kmac_entropy.sv
Comment (vogelpi) [reply hw/ip/kmac/rtl/kmac_entropy.sv:335]: Good point @tjaychen !

File: hw/ip/kmac/rtl/keccak_round.sv
Comment (tjaychen) [reply hw/ip/kmac/rtl/keccak_round.sv:258]: oook got it. So I guess if you didn't have this, you would end up pairing fresh entropy with stale data, which you also do not want, is that right?

**Comment by vogelpi** [comment]
> thanks @vogelpi ! this looks really impressive! I think I generally understand what each commit is trying to do, the main part I'm still not following (and this is more of a general confusing) is when entropy is supplied vs when data is supplied (same question I raised above).
> 
> Otherwise this looks good to me.
> 
> Last note, I almost think all your commit messages are worth becoming its own doc, explaining the general principles of hardening ( a lot of the intermediate re-use and PRNG exercise I believe is also common with AES). What do you think?

Thanks @tjaychen ! The plan is to give a presentation to the security working group on this topic to explain in more depth what's happening (also with some nice figures/diagrams), present results and the methodology developed/used. We could then condense this information into a separate SCA hardening doc or maybe extend @cdgori 's guidance doc.

File: hw/ip/kmac/rtl/keccak_round.sv
Comment (vogelpi) [reply hw/ip/kmac/rtl/keccak_round.sv:258]: Yes exactly. :+1:

File: hw/ip/kmac/rtl/kmac.sv
Comment (eunchan) [inline hw/ip/kmac/rtl/kmac.sv:1151]: We do not need this. entropy module checks EDN latency (`EdnWaitTimer`).

**Review by eunchan — APPROVED** [review]
Overall Looks good to me. Thanks for the work Pirmin.

**Comment by cdgori** [comment]
> Thanks @tjaychen ! The plan is to give a presentation to the security working group on this topic to explain in more depth what's happening (also with some nice figures/diagrams), present results and the methodology developed/used. We could then condense this information into a separate SCA hardening doc or maybe extend @cdgori 's guidance doc.

I think a separate SCA hardening doc would be a good idea, we can point to it in the general guidance doc.

I think of the general guidance as being applicable to almost all of the modules (some non-security exceptions), but SCA hardening mostly/only applies to implementation of crypto primitives, which is a bit more esoteric/specialized situation.  Nothing else should be doing hardward arithmetic operations on keys/secret data.

**Comment by cdgori** [comment]
> I would assume masking of the Keccak core is not strictly required for the part after absorbing the key. However, we did not systematically analyze this. Note that in addition to the masking inside the Keccak core, KMAC also features the option to mask the key/data upon loading into the Keccak state. This was always enabled in our experiments.
> 
> Anyway, with this PR the only reason to disable the masking of the Keccak core (fast_entropy mode) would be to save power. But performance-wise there is no longer a difference.

I tend to agree with this.  Note, that in theory the subsequent KMAC operations could leak information correlated to the internal Keccak state after the key was absorbed.  However, the nice thing about using KMAC (or HMAC, by analogy) for this KDF-type of thing, as opposed to AES, is that by nature of the underlying primitive being a (one-way) hash function, it's nigh-onto-impossible to reverse from the Keccak state to the input key.  Whereas, AES, depending on the exact usage of the primitive, might have issues because it's a symmetric cipher, designed to be used for encryption&decryption and therefore reversible.

The main theoretical attack(s) remaining require the design to leak enough information about the Keccak state while processing the subsequent data such that the key-absorbed state could be fully recovered, and then either: a) use that state offline to forge later/derived identities (unclear how viable this is to me) or fraudulently sign messages, or b) somehow use DFT to scan that state back into a functioning device to create a clone of a device while still not actually knowing the underlying secret (the key that went into the KMAC).

**Review by cdgori — APPROVED** [review]
LGTM, just a few nits.

The explanations of what was done are great, and the result is excellent.

File: hw/ip/kmac/rtl/kmac_pkg.sv
Comment (vogelpi) [reply hw/ip/kmac/rtl/kmac_pkg.sv:100]: I've accidentally used the same seed that was already used by the `RndCnstMsgPermDefault` below (copy & paste of the entire command). I've now switched to a different seed.

File: hw/ip/kmac/rtl/kmac_pkg.sv
Comment (vogelpi) [reply hw/ip/kmac/rtl/kmac_pkg.sv:145]: Good catch. I've now also switched the seed here.

File: hw/ip/kmac/rtl/keccak_2share.sv
Comment (vogelpi) [reply hw/ip/kmac/rtl/keccak_2share.sv:312]: Ah sorry, this was clearly wrong. I think I added this comment when adding the pipeline stage but forgot to revise it after I had understood how things really work. I've now changed that.

Basically, we forward the generated lane half to both the upper and lower lane halves of `chi_data` here. Then we apply Iota to both lane halves to generate `iota_data`. But in the cycle we write the lower lane halves of the state register, the upper lane halves of `iota_data` are ignored (this part also contains the lower lane halves of `chi_data` XOR upper lane half of the round constant - this is what the Iota step adds). Whereas when writing the upper lane halves, the lower lane halves of `iota_data` are ignored (now containing the upper lane halves of `chi_data` XOR the lower lane half of the round constant). There is no logic that synthesis needs to optimize away.

File: hw/ip/kmac/fpv/tb/keccak_2share_fpv.sv
Comment (vogelpi) [reply hw/ip/kmac/fpv/tb/keccak_2share_fpv.sv:15]: I don't think so because this input is also using for masking the input state in this FPV testbench which has a width of `Width` bits.

File: hw/ip/kmac/data/kmac.hjson
Comment (vogelpi) [reply hw/ip/kmac/data/kmac.hjson:660]: Yes indeed, it's a multireg. I've now reworked the description, also to explain that software can write these registers only once and that it should do that in ascending order.

File: hw/ip/kmac/rtl/kmac_entropy.sv
Comment (vogelpi) [reply hw/ip/kmac/rtl/kmac_entropy.sv:323]: They way I understood it, if configured to use EDN software can trigger an entropy refresh through the command register whenever needed. This is from `kmac.CMD @ 0x18`:

Screenshot from 2022-04-22 15-55-45

But adding an extra hook in hardware to trigger five reseeds would certainly be possible.

File: hw/ip/kmac/rtl/kmac_entropy.sv
Comment (vogelpi) [reply hw/ip/kmac/rtl/kmac_entropy.sv:335]: This is still pending, but I would like to implement a prim_count for this.

**Review by vogelpi — COMMENTED** [review]
Thanks everybody for your reviews! I've now implemented most of it. What's still left is the documentation + adding the prim_count for the reseeding counter. Will continue with this later today.

File: hw/ip/kmac/doc/_index.md
Comment (vogelpi) [reply hw/ip/kmac/doc/_index.md]: I've now also aligned the documentation.

File: hw/ip/kmac/rtl/kmac_entropy.sv
Comment (vogelpi) [reply hw/ip/kmac/rtl/kmac_entropy.sv:356]: I've now switched to a `prim_count` for this.

File: hw/ip/kmac/doc/_index.md
Comment (vogelpi) [inline hw/ip/kmac/doc/_index.md:22]: @eunchan , I wasn't sure how the throughput for SHA3 should be computed based on the number of rounds and cycles per rounds. What I did now is just scale down by 33% as a single round now takes 4 instead of 3 clock cycles, but I don't know if that's accurate. Do you have a formula for this somewhere?

**Comment by vogelpi** [comment]
Update: I've now also done the documentation and switched to a prim_count for the seed index counter as recommended by Tim. I would say this is ready to be merged.

**Comment by vogelpi** [comment]
Update: I've done another rebase and will merge this when CI passes.

File: hw/ip/kmac/rtl/kmac.sv
Comment (a-will) [inline hw/ip/kmac/rtl/kmac.sv:1108]: Note: This violates the `prim_sync_reqack` interface. To use `prim_sync_reqack`, entropy_req cannot be allowed to time out, else it will lose synchronization.

Comment (vogelpi) [reply hw/ip/kmac/rtl/kmac.sv:1108]: Yes, I think we should simply add a flop in the KMAC clock domain to always keep `entropy_req` asserted until the next `ack` independent of potential timeouts. Let's discuss in @matutem 's issue.

Comment (vogelpi) [reply hw/ip/kmac/rtl/kmac.sv:1108]: FYI: the assertion got dynamically disabled later on in 2f044ef0e564a055f1a0d639eb646fc7af7be606 as follows:
```sv
  .req_chk_i ((kmac_entropy_state_error == 1'b0) && (entropy_err.valid == 1'b0)),
```
But this didn't work because the valid bit is high just during a single clock cycle. Maybe for that reason the SVA got completely disabled for the corresponding block-level test in the same commit. Of course this doesn't work for top-level DV. Anyway, I'll file a PR to hopefully fix this for good in a second.

**Pull request merged**
